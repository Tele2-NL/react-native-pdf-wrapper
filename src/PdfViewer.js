import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Platform, NativeModules, ActivityIndicator, View, Text, StyleSheet } from 'react-native';
import Pdf from 'react-native-pdf';
import PdfWebView from './PdfWebView';

const { PdfViewManager } = NativeModules;

const IS_IOS = Platform.OS === 'ios';

class PdfViewer extends Component {
  static propTypes = {
    ...Pdf.propTypes,
    loading: PropTypes.node,
    error: PropTypes.node,
  };

  static defaultProps = {
    onLoadComplete: () => {},
    onError: () => {},
    loading: <ActivityIndicator />,
    error: <Text>Error thing</Text>,
  };

  state = {
    isLoading: true,
    hasError: false,
    supportsPDFKit: null,
  };

  componentDidMount() {
    if (IS_IOS) {
      PdfViewManager.supportPDFKit(this.handleSupportPDFKit);
    }
  }

  handleSupportPDFKit = supportsPDFKit =>
    this.setState({
      supportsPDFKit,
    });

  handlePdfLoaded = () => {
    this.setState({ isLoading: false });

    this.props.onLoadComplete();
  };

  handleError = (error) => {
    this.setState({
      isLoading: false,
      hasError: true,
    });

    this.props.onError(error);
  };

  renderError = () => React.cloneElement(this.props.error);

  renderLoading = () => React.cloneElement(this.props.loading);

  renderPdf = () => {
    const { hasError, supportsPDFKit } = this.state;

    if (hasError) {
      return this.renderError();
    }

    if (supportsPDFKit === null) {
      return null;
    }

    if (!supportsPDFKit) {
      return (
        <PdfWebView
          {...this.props}
          onLoadComplete={this.handlePdfLoaded}
          onError={this.handleError}
        />
      );
    }

    return (
      <Pdf
        {...this.props}
        onLoadComplete={this.handlePdfLoaded}
        activityIndicator={<View />}
        onError={this.handleError}
        style={styles.pdf}
      />
    );
  }

  render() {
    const { Loading, ...props } = this.props;
    const { isLoading, supportsPDFKit } = this.state;

    return (
      <View style={styles.container}>
        {isLoading && this.renderLoading()}

        {this.renderPdf()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  pdf: {
    flex: 1,
    width: '100%',
  },
});

export default PdfViewer;
