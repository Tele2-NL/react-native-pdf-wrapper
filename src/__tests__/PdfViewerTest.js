import React from 'react';
import { Platform, NativeModules } from 'react-native';
import { shallow } from 'enzyme';
import PdfViewer from '../PdfViewer';

describe('<PdfViewer />', () => {
  beforeEach(() => {
    this.props = {
      source: {
        uri: 'https://remote/path/to/file',
      },
      onError: jest.fn(),
    };

    NativeModules.PdfViewManager.supportPDFKit = jest.fn(cb => cb(true));
    NativeModules.PdfViewManager.supportPDFKit.mockClear();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<PdfViewer {...this.props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should check if PDFKit is supported if platform is iOS', () => {
    Platform.OS = 'ios';

    shallow(<PdfViewer {...this.props} />);

    expect(NativeModules.PdfViewManager.supportPDFKit).toHaveBeenCalledTimes(1);
  });

  it('should store in state if PDFKit is supported', () => {
    Platform.OS = 'ios';

    const wrapper = shallow(<PdfViewer {...this.props} />);

    expect(wrapper.state().supportsPDFKit).toBe(true);
  });

  it('should fallback to `PdfWebView` if PDFKit is not supported', () => {
    Platform.OS = 'ios';

    NativeModules.PdfViewManager.supportPDFKit = jest.fn(cb => cb(false));

    const wrapper = shallow(<PdfViewer {...this.props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call `onError` if an error occurs', () => {
    const error = 'error';

    const wrapper = shallow(<PdfViewer {...this.props} />);

    wrapper.instance().handleError(error);

    expect(this.props.onError).toHaveBeenCalledWith(error);
  });

  it('should store that an error happened if an error occurs', () => {
    const error = 'error';

    const wrapper = shallow(<PdfViewer {...this.props} />);

    wrapper.instance().handleError(error);

    expect(wrapper.state().isLoading).toBe(false);
    expect(wrapper.state().hasError).toBe(true);
  });
});
