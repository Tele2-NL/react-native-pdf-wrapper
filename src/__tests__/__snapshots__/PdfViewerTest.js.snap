// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PdfViewer /> renders correctly 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Component />
  <Pdf
    activityIndicator={<Component />}
    activityIndicatorProps={
      Object {
        "color": "#009900",
        "progressTintColor": "#009900",
      }
    }
    enablePaging={false}
    enableRTL={false}
    error={
      <Error
        retryPdf={[Function]}
      />
    }
    fitPolicy={2}
    horizontal={false}
    loading={<Component />}
    onError={[Function]}
    onLoadComplete={[Function]}
    onLoadProgress={[Function]}
    onPageChanged={[Function]}
    onPageSingleTap={[Function]}
    onScaleChanged={[Function]}
    page={1}
    password=""
    scale={1}
    source={
      Object {
        "uri": "https://remote/path/to/file",
      }
    }
    spacing={10}
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
  />
</Component>
`;

exports[`<PdfViewer /> should fallback to \`PdfWebView\` if PDFKit is not supported 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Component />
  <PdfWebView
    error={
      <Error
        retryPdf={[Function]}
      />
    }
    loading={<Component />}
    onError={[Function]}
    onLoadComplete={[Function]}
    source={
      Object {
        "uri": "https://remote/path/to/file",
      }
    }
  />
</Component>
`;
